tracer/src/Datadog.Trace/Headers/CarrierWithDelegate.cs

```diff
@@ -0,0 +1,67 @@
+// <copyright file="CarrierWithDelegate.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Datadog.Trace.Headers;
+
+/// <summary>
+/// Builds an IHeadersCollection from a carrier and a delegate to get or set data.
+/// </summary>
+internal readonly struct CarrierWithDelegate<TCarrier> : IHeadersCollection
```

Comment:
I created this class because DSM methods take an IHeadersCollection instead of a delegate, but I could also write overloads taking a delegate instead. The call chain can be a bit long though, so I chose this approach to minimize the amount of code I needed to change.

Reply:
FYI: This of defeat the purpose of the struct + generics constraints optimization in the propagator parsers. That way enables two features. First allow the jit to remove unused branches and to apply optimizations in inlining, and second avoid any closure, boxing and others `unseen` allocations in a hot path.

Reply:
ok, I understand, though this is meant to be applied in the context of a manual API call by a user, where the current API already takes a delegate.
I imagine that when users call us themselves, we can value ease of use a little bit above raw perf ?

Reply:
Ok if is a public api usage then it makes sense.

Reply:
I amended the comment to make that more explicit